packages/field-base/src/error-controller.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * @license
+ * Copyright (c) 2021 - 2022 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+import { SlotController } from '@vaadin/component-base/src/slot-controller.js';
+
+/**
+ * A controller that manages the error message node content.
+ */
+export class ErrorController extends SlotController {
+  constructor(host) {
+    super(
+      host,
+      'error-message',
+      () => document.createElement('div'),
+      (_host, node) => {
+        this.__updateErrorId(node);
+
+        this.__updateHasError();
+      }
+    );
+  }
+
+  /**
+   * ID attribute value set on the error message element.
+   *
+   * @return {string}
+   */
+  get errorId() {
+    return this.node && this.node.id;
+  }
+
+  /**
+   * Set the error message element text content.
+   *
+   * @param {string} errorMessage
+   */
+  setErrorMessage(errorMessage) {
+    this.errorMessage = errorMessage;
+
+    this.__updateHasError();
+  }
+
+  /**
+   * Set invalid state for detecting whether to show error message.
+   *
+   * @param {boolean} invalid
+   */
+  setInvalid(invalid) {
+    this.invalid = invalid;
+
+    this.__updateHasError();
+  }
+
+  /**
+   * Override to initialize the newly added custom label.
+   *
+   * @param {Node} errorNode
+   * @protected
+   * @override
+   */
+  initCustomNode(errorNode) {
+    this.__updateErrorId(errorNode);
+
+    // Save the custom error message content on the host.
+    if (errorNode.textContent && !this.errorMessage) {
+      this.host.errorMessage = errorNode.textContent.trim();
```

Comment:
We actually need to update error message on the host (there is a test for it).
Then it gets updated on the controller as well through the observer.

Reply:
I'm not generally very happy about this approach because it means that the controller implicitly depends on a certain behavior of the host component, the contract is not obvious. In more technical details, the current controller isn't entirely independent: it has a `setErrorMessage` method but at the same time, it relies on a `host.errorMessage` property and assumes there is an observer. Although tests enforce this behavior, this kind of implicit relation can be more difficult to figure out/maintain in the future.

Maybe the right way would be to fire an event that can be handled by the host component to synchronize the `errorMessage` property or whatever. The HelperController and LabelController don't sync their corresponding host properties though, shouldn't the same goes to the ErrorController?

Reply:
Updated, also changed to not return slotted error message as `errorMessage` (same as for helper).